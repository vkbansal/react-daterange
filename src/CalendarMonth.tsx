import addDays from 'date-fns/addDays';
import isAfter from 'date-fns/isAfter';
import isBefore from 'date-fns/isBefore';
import isSameDay from 'date-fns/isSameDay';
import isSameMonth from 'date-fns/isSameMonth';
import setMonth from 'date-fns/setMonth';
// import setYear from 'date-fns/setYear';
import startOfMonth from 'date-fns/startOfMonth';
import glamorous, { CSSProperties } from 'glamorous';
import range from 'lodash.range';
import * as React from 'react';

import { NavButton } from './Common';

const SIZE = 38;
const sizeInPixels = `${SIZE}px`;

export const Month = glamorous('div')({
    display: 'inline-block',
    padding: '0 16px',
    background: '#fff',
    minHeight: `${SIZE * 8}px`
});

Month.displayName = 'Month';

export interface RowProps {
    justifyContent?: CSSProperties['justifyContent'];
}

export const Row = glamorous('div')<RowProps>(
    {
        display: 'flex'
    },
    props => ({
        justifyContent: props.justifyContent || 'flex-start'
    })
);

Row.displayName = 'Row';

interface CellProps {
    span?: number;
    justifyContent?: CSSProperties['justifyContent'];
}

export const Cell = glamorous('div')<CellProps>(
    {
        height: sizeInPixels,
        lineHeight: sizeInPixels,
        fontWeight: 'bold',
        fontSize: '14px',
        textAlign: 'center',
        display: 'flex'
    },
    props => ({
        width: `${SIZE * (props.span || 1)}px`,
        justifyContent: props.justifyContent || 'center'
    })
);

Cell.displayName = 'Cell';

interface DayProps {
    selected?: boolean;
    inRange?: boolean;
    isDisabled?: boolean;
}

export const Day = glamorous(Cell)<DayProps>(
    {
        width: `${SIZE + 1}px`,
        height: `${SIZE + 1}px`,
        lineHeight: `${SIZE + 1}px`,
        fontWeight: 'normal',
        border: '1px solid #ddd',
        marginTop: '-1px',
        marginLeft: '-1px',
        '&:last-child': {
            borderRight: '1px solid #ddd'
        }
    },
    props => ({
        cursor: props.isDisabled ? 'not-allowed' : 'pointer',
        textDecoration: props.isDisabled ? 'line-through' : 'none',
        color: props.isDisabled ? '#ddd' : props.selected ? '#fff' : '#222',
        background: props.selected
            ? 'rgb(0, 202, 255)'
            : props.inRange ? 'rgba(0, 202, 255, 0.3)' : 'transparent',
        '&:hover': {
            background: props.isDisabled
                ? 'transparent'
                : props.selected ? 'rgb(0, 202, 255)' : '#ddd'
        }
    })
);

Day.displayName = 'Day';

export const Select = glamorous('select')({
    background: 'transparent',
    border: 'none',
    outline: 'none'
});

Select.displayName = 'Select';

export interface CalendarMonthProps {
    month: Date;
    startDate?: Date;
    endDate?: Date;
    minDate?: Date;
    maxDate?: Date;
    showDropdowns?: boolean;
    daysOfWeek?: Array<string>;
    monthNames?: Array<string>;
    onPrevClick?: () => void;
    onNextClick?: () => void;
    onDayClick?: (date: Date) => void;
    onDayHover?: (date: Date) => void;
    onMonthChange?: (month: number) => void;
    onYearChange?: (year: number) => void;
    hideNextButton?: boolean;
    hidePrevButton?: boolean;
}

export class CalendarMonth extends React.Component<CalendarMonthProps> {
    static defaultProps = {
        daysOfWeek: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
        monthNames: [
            'January',
            'February',
            'March',
            'April',
            'May',
            'June',
            'July',
            'August',
            'September',
            'October',
            'November',
            'December'
        ]
    };

    constructor(props: CalendarMonthProps) {
        super(props);

        if (!Array.isArray(props.daysOfWeek) || props.daysOfWeek.length !== 7) {
            throw new Error(
                `props.daysofWeek must an array of length 7 but ${props.daysOfWeek} given`
            );
        }

        if (!Array.isArray(props.monthNames) || props.monthNames.length !== 12) {
            throw new Error(
                `props.monthNames must an array of length 7 but ${props.daysOfWeek} given`
            );
        }
    }

    handlePrevClick = (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();

        if (typeof this.props.onPrevClick === 'function') {
            this.props.onPrevClick();
        }
    };

    handleNextClick = (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();

        if (typeof this.props.onNextClick === 'function') {
            this.props.onNextClick();
        }
    };

    handleDayClick = (day: Date) => (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();

        if (typeof this.props.onDayClick === 'function') {
            this.props.onDayClick(day);
        }
    };

    handleDayHover = (day: Date) => (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();

        if (typeof this.props.onDayHover === 'function') {
            this.props.onDayHover(day);
        }
    };

    handleMonthChange = (e: React.SyntheticEvent<EventTarget>) => {
        const { value } = e.target as HTMLOptionElement;
        if (typeof this.props.onMonthChange === 'function') {
            this.props.onMonthChange(parseInt(value, 10));
        }
    };

    handleYearChange = (e: React.SyntheticEvent<EventTarget>) => {
        const { value } = e.target as HTMLOptionElement;
        if (typeof this.props.onYearChange === 'function') {
            this.props.onYearChange(parseInt(value, 10));
        }
    };

    renderDropDowns = (firstDay: Date) => {
        const { monthNames, minDate, maxDate, month } = this.props;

        return [
            <Select key="months" value={firstDay.getMonth()} onChange={this.handleMonthChange}>
                {(monthNames || []).map((m, i) => {
                    const isDisabled =
                        (minDate && isBefore(setMonth(month, i), minDate)) ||
                        (maxDate && isAfter(setMonth(month, i), maxDate));

                    return (
                        <option key={i} value={i} disabled={isDisabled}>
                            {m}
                        </option>
                    );
                })}
            </Select>,
            <Select key="years" value={firstDay.getFullYear()} onChange={this.handleYearChange}>
                {range(
                    minDate ? minDate.getFullYear() : firstDay.getFullYear() - 50,
                    maxDate ? maxDate.getFullYear() + 1 : firstDay.getFullYear() + 5
                ).map(i => {
                    return (
                        <option key={i} value={i}>
                            {i}
                        </option>
                    );
                })}
            </Select>
        ];
    };

    render() {
        const {
            month,
            daysOfWeek,
            monthNames,
            startDate,
            endDate,
            hideNextButton,
            hidePrevButton,
            minDate,
            maxDate,
            showDropdowns
        } = this.props;

        if (!daysOfWeek || !monthNames) return null;

        const firstDay: Date = startOfMonth(month);
        const firstDayInWeek: number = firstDay.getDay();
        const monthName = monthNames[firstDay.getMonth()];
        const key = `${monthName}-${firstDay.getFullYear()}`;

        return (
            <Month>
                <Row>
                    <Cell>
                        {!hidePrevButton && (
                            <NavButton onClick={this.handlePrevClick}>&#8592;</NavButton>
                        )}
                    </Cell>
                    <Cell span={5} justifyContent={showDropdowns ? 'space-around' : 'center'}>
                        {showDropdowns
                            ? this.renderDropDowns(firstDay)
                            : `${monthName} ${firstDay.getFullYear()}`}
                    </Cell>
                    <Cell>
                        {!hideNextButton && (
                            <NavButton onClick={this.handleNextClick}>&#8594;</NavButton>
                        )}
                    </Cell>
                </Row>
                <Row>
                    {range(0, 7).map(i => <Cell key={`${key}-day-${i}`}>{daysOfWeek[i]}</Cell>)}
                </Row>
                {range(0, 6).map(i => (
                    <Row key={`${key}-${i}`} justifyContent={i === 0 ? 'flex-end' : 'flex-start'}>
                        {range(0, 7).map(j => {
                            const day = addDays(firstDay, i * 7 + j - firstDayInWeek);

                            if (!isSameMonth(firstDay, day)) return null;

                            const inRange =
                                startDate &&
                                isAfter(day, startDate) &&
                                endDate &&
                                isBefore(day, endDate);

                            const selected =
                                (startDate && isSameDay(day, startDate)) ||
                                (endDate && isSameDay(day, endDate));
                            const isDisabled =
                                (minDate && isBefore(day, minDate)) ||
                                (maxDate && isAfter(day, maxDate));

                            return (
                                <Day
                                    key={`${key}-${i}-${j}`}
                                    selected={selected}
                                    inRange={inRange}
                                    onClick={!isDisabled ? this.handleDayClick(day) : undefined}
                                    onMouseEnter={
                                        !isDisabled ? this.handleDayHover(day) : undefined
                                    }
                                    isDisabled={isDisabled}
                                >
                                    {day.getDate()}
                                </Day>
                            );
                        })}
                    </Row>
                ))}
            </Month>
        );
    }
}

/**
 * TODO:
 * - add & handle maxDropdownYear prop
 * - add & handle maxDropdownMonth prop
 */
