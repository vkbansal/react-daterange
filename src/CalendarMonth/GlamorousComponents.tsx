import * as React from 'react';
import glamorous, { GlamorousComponent, ExtraGlamorousProps } from 'glamorous';

import { getStyleOverrides } from '../utils/glamorousUtils';

export { GlamorousComponent, React, ExtraGlamorousProps }; // Required, just for generating type definitions

export const Month = glamorous.div(
    {
        display: 'inline-flex',
        alignItems: 'flex-start',
        flexFlow: 'column',
        padding: '0 16px',
        minHeight: '304px'
    },
    getStyleOverrides('month')
);

Month.displayName = 'Month';

export interface CellProps {
    span?: number;
}

export const Cell = glamorous.div<CellProps>(
    {
        display: 'flex',
        height: '38px',
        lineHeight: '38px',
        fontWeight: 'bold',
        fontSize: '14px',
        alignItems: 'center',
        textAlign: 'center',
        justifyContent: 'center'
    },
    (props) => ({
        width: `${(props.span || 1) * 38}px`
    }),
    getStyleOverrides<CellProps>('cell')
);

export interface DayProps {
    selected?: boolean;
    disabled?: boolean;
    inRange?: boolean;
    today?: boolean;
}

export const Day = glamorous.div<DayProps>(
    {
        height: '39px',
        width: '39px',
        lineHeight: '39px',
        margin: '-1px 0 0 -1px',
        fontWeight: 'normal',
        border: '1px solid #ddd',
        cursor: 'pointer',
        background: 'transparent',
        textAlign: 'center',
        fontSize: '14px',
        transition: 'background 0.2s ease-out'
    },
    (props) => ({
        background: (() => {
            if (props.selected) {
                return 'rgba(0, 202, 255, 1)';
            }

            if (props.today) {
                return '#e94949';
            }

            if (props.inRange) {
                return 'rgba(0, 202, 255, 0.3)';
            }

            return 'transparent';
        })(),
        color: (() => {
            if (props.selected || props.today) {
                return '#fff';
            }

            if (props.disabled) {
                return '#ddd';
            }

            return '#222';
        })(),
        cursor: props.disabled ? 'not-allowed' : 'pointer',
        textDecoration: props.disabled ? 'line-through' : 'none',
        '&:hover': {
            background: (() => {
                if (props.disabled) {
                    return 'transparent';
                }

                if (props.selected) {
                    return 'rgba(0, 202, 255, 1)';
                }

                if (props.inRange) {
                    return 'rgba(0, 202, 255, 0.3)';
                }

                return '#ddd';
            })(),
            color: (() => {
                if (props.selected) {
                    return '#fff';
                }

                if (props.disabled) {
                    return '#ddd';
                }

                return '#222';
            })()
        }
    }),
    getStyleOverrides<DayProps>('day')
);

Day.displayName = 'Day';

export const Row = glamorous.div(
    {
        display: 'flex',
        justifyContent: 'flex-start',
        width: '100%'
    },
    getStyleOverrides('row')
);

Row.displayName = 'Row';

export const Weeks = glamorous.div(
    {
        height: '240px'
    },
    getStyleOverrides('weeks')
);

Weeks.displayName = 'Weeks';

export const Week = glamorous.div(
    {
        display: 'flex',
        justifyContent: 'flex-start',
        width: '100%',
        '&:first-child': {
            justifyContent: 'flex-end'
        }
    },
    getStyleOverrides('week')
);

Week.displayName = 'Week';

export const NavButton = glamorous.button(
    {
        background: 'transparent',
        border: 'none',
        fontWeight: 'bold',
        fontSize: '18px',
        outline: 'none',
        cursor: 'pointer',
        width: '42px',
        height: '42px',
        lineHeight: '42px'
    },
    getStyleOverrides('navButton')
);

NavButton.displayName = 'NavButton';

export const Select = glamorous.select(
    {
        background: 'transparent',
        border: '1px solid #fff',
        outline: 'none',
        boxShadow: 'none',
        '& + &': {
            marginLeft: '4px'
        }
    },
    getStyleOverrides('select')
);

Select.displayName = 'Select';
