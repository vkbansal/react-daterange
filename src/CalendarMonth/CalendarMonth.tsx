/* tslint:disable:max-line-length */
import * as React from 'react';

import { Day, Row, Cell, NavButton, Weeks, Week, Select, Month } from './GlamorousComponents';
import {
    LOCALE_EN,
    addDays,
    getWeeksInMonth,
    isDayAfter,
    isDayBefore,
    isSameDay,
    isSameMonth,
    setMonth,
    startOfMonth
} from '../utils/dateUtils';
import { callIfExists } from '../utils/otherUtils';

export interface CalendarMonthProps {
    /**
     * Current Month
     */
    month: Date;
    /**
     * The start of the initially selected date range
     */
    startDate?: Date;
    /**
     * The end of the initially selected date range
     */
    endDate?: Date;
    /**
     * The earliest date a user may select
     */
    minDate?: Date;
    /**
     * The latest date a user may select
     */
    maxDate?: Date;
    /**
     * An array of strings, that is used to display the days of the week in the calendar.
     * Note: first day in the array must be a sunday
     * @default ['Sun','Mon','Tue','Wed','Thu','Fri','Sat']
     */
    daysOfWeek?: string[];
    /**
     *  An array of strings, that is used to display the month names in the calendar.
     * @default ['January','February','March','April','May','June','July','August','September','October','November','December']
     */
    monthNames?: string[];
    /**
     * If set as `true`, shows **month** and **year** dropdowns above the calendar
     */
    showDropdowns?: boolean;
    /**
     * Callback for when "prev" button is clicked
     */
    onPrevClick?(): void;
    /**
     * Callback for when "next" button is clicked
     */
    onNextClick?(): void;
    /**
     * Callback for when user clicks on a "Day"
     */
    onDayClick?(date: Date): void;
    /**
     * Callback for when user hovers on a "Day"
     */
    onDayHover?(date: Date): void;
    /**
     * Callback for when user selects a month from the dropdown
     * Applicable only when `showDropdowns: true`
     */
    onMonthChange?(month: number): void;
    /**
     * Callback for when user selects a year from the dropdown
     * Applicable only when `showDropdowns: true`
     */
    onYearChange?(year: number): void;
    /**
     * When sets as `true`, hide the "next" button
     */
    hideNextButton?: boolean;
    /**
     * When sets as `true`, hide the "prev" button
     */
    hidePrevButton?: boolean;
    /**
     * TODO: implement this feature
     */
    showWeekNumbers?: boolean;
    /**
     * When enabled, week starts on monday. See [ISO week date](https://en.wikipedia.org/wiki/ISO_week_date)
     */
    showISOWeek?: boolean;
}

export class CalendarMonth extends React.Component<CalendarMonthProps> {
    handlePrevClick = (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();

        if (typeof this.props.onPrevClick === 'function') {
            this.props.onPrevClick();
        }
    };

    handleNextClick = (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();
        callIfExists(this.props.onNextClick);
    };

    handleDayClick = (day: Date) => (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();
        callIfExists(this.props.onDayClick, day);
    };

    handleDayHover = (day: Date) => (e: React.SyntheticEvent<EventTarget>) => {
        e.preventDefault();
        callIfExists(this.props.onDayHover, day);
    };

    handleMonthChange = (e: React.FormEvent<HTMLSelectElement>) => {
        const { value } = e.target as HTMLSelectElement;
        callIfExists(this.props.onMonthChange, parseInt(value, 10));
    };

    handleYearChange = (e: React.FormEvent<HTMLSelectElement>) => {
        const { value } = e.target as HTMLSelectElement;
        callIfExists(this.props.onYearChange, parseInt(value, 10));
    };

    renderDropDowns = (firstDay: Date) => {
        const { minDate, maxDate, month, monthNames } = this.props;

        const localeMonthNames = monthNames || LOCALE_EN.monthNames;
        const today = new Date();
        const minYear = minDate ? minDate.getFullYear() : today.getFullYear() - 50;
        const maxYear = maxDate ? maxDate.getFullYear() + 1 : today.getFullYear() + 5;

        return (
            <>
                <Select value={firstDay.getMonth()} onChange={this.handleMonthChange}>
                    {localeMonthNames.map((m, i) => {
                        const isDisabled =
                            (minDate && isDayBefore(setMonth(month, i), minDate)) ||
                            (maxDate && isDayAfter(setMonth(month, i), maxDate));

                        return (
                            <option key={i} value={i} disabled={isDisabled}>
                                {m}
                            </option>
                        );
                    })}
                </Select>
                <Select value={firstDay.getFullYear()} onChange={this.handleYearChange}>
                    {Array.from({ length: maxYear - minYear }, (_, i) => {
                        const year = i + minYear;

                        return (
                            <option key={year} value={year}>
                                {year}
                            </option>
                        );
                    })}
                </Select>
            </>
        );
    };

    render() {
        const {
            month,
            startDate,
            endDate,
            hideNextButton,
            hidePrevButton,
            minDate,
            maxDate,
            daysOfWeek,
            monthNames,
            showDropdowns,
            showISOWeek
            // showWeekNumbers
        } = this.props;

        /**
         * 1st of the month
         */
        const firstDate = startOfMonth(month);

        /**
         * Day of week of 1st of the month
         * Sunday = 0
         * Monday = 1
         * Tuesday = 2
         * Wednesday = 3
         * Thursday = 4
         * Friday = 5
         * Saturday = 6
         */
        let firstDayInCurrentMonth = firstDate.getDay();

        /**
         * Since, ISOWeek starts on a Monday and `weeksOfDays` starts on Sunday,
         * we shift the days to left by 1 day
         * Monday = 0
         * Tuesday = 1
         * Wednesday = 2
         * Thursday = 3
         * Friday = 4
         * Saturday = 5
         * Sunday = 6
         */
        if (showISOWeek) {
            firstDayInCurrentMonth = (firstDayInCurrentMonth + 6) % 7;
        }

        const monthName = (monthNames || LOCALE_EN.monthNames)[firstDate.getMonth()];
        const key = `${monthName}-${firstDate.getFullYear()}`;
        const localeDaysOfWeek = daysOfWeek || LOCALE_EN.daysOfWeek;
        const weeks = getWeeksInMonth(firstDate, showISOWeek);
        const today = new Date();

        return (
            <Month>
                <Row>
                    <Cell>
                        {!hidePrevButton && (
                            <NavButton onClick={this.handlePrevClick}>&#8592;</NavButton>
                        )}
                    </Cell>
                    <Cell span={5}>
                        {showDropdowns
                            ? this.renderDropDowns(firstDate)
                            : `${monthName} ${firstDate.getFullYear()}`}
                    </Cell>
                    <Cell>
                        {!hideNextButton && (
                            <NavButton onClick={this.handleNextClick}>&#8594;</NavButton>
                        )}
                    </Cell>
                </Row>
                <Row>
                    {Array.from({ length: 7 }, (_, i) => (
                        <Cell key={`${key}-day-${i}`}>
                            {localeDaysOfWeek[showISOWeek ? (i + 1) % 7 : i]}
                        </Cell>
                    ))}
                </Row>
                <Weeks>
                    {Array.from({ length: weeks }, (_, i) => (
                        <Week key={`${key}-${i}`}>
                            {Array.from({ length: 7 }, (_1, j) => {
                                const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                                if (!isSameMonth(firstDate, day)) return null;

                                const inRange =
                                    startDate &&
                                    isDayAfter(day, startDate) &&
                                    endDate &&
                                    isDayBefore(day, endDate);

                                const selected =
                                    (startDate && isSameDay(day, startDate)) ||
                                    (endDate && isSameDay(day, endDate));
                                const isDisabled =
                                    (minDate && isDayBefore(day, minDate)) ||
                                    (maxDate && isDayAfter(day, maxDate));

                                return (
                                    <Day
                                        selected={selected}
                                        inRange={inRange}
                                        today={isSameDay(today, day)}
                                        disabled={isDisabled}
                                        key={`${key}-${i}-${j}`}
                                        onClick={!isDisabled ? this.handleDayClick(day) : undefined}
                                        onMouseEnter={
                                            !isDisabled ? this.handleDayHover(day) : undefined
                                        }
                                    >
                                        {day.getDate()}
                                    </Day>
                                );
                            })}
                        </Week>
                    ))}
                </Weeks>
            </Month>
        );
    }
}
