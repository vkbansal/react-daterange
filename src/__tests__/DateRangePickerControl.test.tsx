import { shallow } from 'enzyme';
import * as React from 'react';
import * as format from 'date-fns/format';

import { DateRangePickerControl } from '../DateRangePickerControl';
import { CalendarMonth } from '../CalendarMonth';

describe('<DateRangePickerControl/>', () => {
    const startDate = new Date(2017, 0 /* Jan */, 15, 0, 0, 0, 0); // 2017-01-15
    const endDate = new Date(2017, 2 /* Mar */, 14, 0, 0, 0, 0); // 2017-03-14

    test('by default renders consecutive months', () => {
        const component = shallow(
            <DateRangePickerControl startDate={startDate} endDate={endDate} />
        );

        let monthLeft = component.find(CalendarMonth).at(0);
        let monthRight = component.find(CalendarMonth).at(1);

        // Navigation Buttons
        expect(monthLeft.prop('hideNextButton')).toBe(true);
        expect(monthRight.prop('hidePrevButton')).toBe(true);

        // months
        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201702');

        expect(component).toMatchSnapshot();

        // click on next arrow
        monthRight.prop('onNextClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201702');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201703');

        // click on next arrow
        monthRight.prop('onNextClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201703');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201704');

        // click on prev arrow
        monthLeft.prop('onPrevClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201702');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201703');
    });

    test('`inidividualCalenders` renders seperate months', () => {
        const component = shallow(
            <DateRangePickerControl individualCalendars startDate={startDate} endDate={endDate} />
        );

        let monthLeft = component.find(CalendarMonth).at(0);
        let monthRight = component.find(CalendarMonth).at(1);

        // Navigation Buttons
        expect(monthLeft.prop('hideNextButton')).toBe(false);
        expect(monthRight.prop('hidePrevButton')).toBe(false);

        // months
        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201703');

        expect(component).toMatchSnapshot();

        // click on next arrow in month on the right
        monthRight.prop('onNextClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201704');

        // click on next arrow in month on the left
        monthLeft.prop('onNextClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201702');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201704');

        // click on next arrow in month on the left
        monthLeft.prop('onNextClick')();
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201703');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201704');
    });

    test('renders consecutive months with dropdowns', () => {
        const component = shallow(
            <DateRangePickerControl showDropdowns startDate={startDate} endDate={endDate} />
        );

        let monthLeft = component.find(CalendarMonth).at(0);
        let monthRight = component.find(CalendarMonth).at(1);

        // Navigation Buttons
        expect(monthLeft.prop('hideNextButton')).toBe(true);
        expect(monthRight.prop('hidePrevButton')).toBe(true);

        // months
        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201702');

        expect(component).toMatchSnapshot();

        // change month value in month on the right
        monthRight.prop('onMonthChange')(10);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201710');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201711');

        // change year value in month on the right
        monthRight.prop('onYearChange')(2018);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201810');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201811');

        // change year value in month on the left
        monthLeft.prop('onYearChange')(2019);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201910');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201911');

        // change month value in month on the left
        monthLeft.prop('onMonthChange')(5);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201906');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201907');
    });

    test('`inidividualCalenders` renders seperate months with dropdowns', () => {
        const component = shallow(
            <DateRangePickerControl
                individualCalendars
                showDropdowns
                startDate={startDate}
                endDate={endDate}
            />
        );

        let monthLeft = component.find(CalendarMonth).at(0);
        let monthRight = component.find(CalendarMonth).at(1);

        // Navigation Buttons
        expect(monthLeft.prop('hideNextButton')).toBe(false);
        expect(monthRight.prop('hidePrevButton')).toBe(false);

        // months
        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201703');

        expect(component).toMatchSnapshot();

        // change month value in month on the right
        monthRight.prop('onMonthChange')(10);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201711');

        // change year value in month on the right
        monthRight.prop('onYearChange')(2018);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201701');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201811');

        // change year value in month on the left
        monthLeft.prop('onYearChange')(2018);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201801');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201811');

        // change month value in month on the left
        monthLeft.prop('onMonthChange')(5);
        component.update();

        monthLeft = component.find(CalendarMonth).at(0);
        monthRight = component.find(CalendarMonth).at(1);

        expect(format(monthLeft.prop('month'), 'YYYYMM')).toBe('201806');
        expect(format(monthRight.prop('month'), 'YYYYMM')).toBe('201811');
    });

    // test('date selection works', () => {
    //     const component = shallow(
    //         <DateRangePickerControl showDropdowns startDate={startDate} endDate={endDate} />
    //     );

    //     let monthLeft = component.find(CalendarMonth).at(0);
    //     let monthRight = component.find(CalendarMonth).at(1);
    // });
});
