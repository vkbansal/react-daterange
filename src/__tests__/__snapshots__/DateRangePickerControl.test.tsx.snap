// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateRangePickerControl/> \`inidividualCalenders\` renders seperate months 1`] = `
<div
  className="rdr-range-control-wrapper"
>
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hideNextButton={false}
    month={2017-01-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    startDate={2017-01-14T18:30:00.000Z}
  />
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hidePrevButton={false}
    minDate={2017-01-31T18:29:59.999Z}
    month={2017-03-13T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    startDate={2017-01-14T18:30:00.000Z}
  />
</div>
`;

exports[`<DateRangePickerControl/> \`inidividualCalenders\` renders seperate months with dropdowns 1`] = `
<div
  className="rdr-range-control-wrapper"
>
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hideNextButton={false}
    month={2017-01-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    showDropdowns={true}
    startDate={2017-01-14T18:30:00.000Z}
  />
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hidePrevButton={false}
    minDate={2017-01-31T18:29:59.999Z}
    month={2017-03-13T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    showDropdowns={true}
    startDate={2017-01-14T18:30:00.000Z}
  />
</div>
`;

exports[`<DateRangePickerControl/> by default renders consecutive months 1`] = `
<div
  className="rdr-range-control-wrapper"
>
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hideNextButton={true}
    month={2017-01-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    startDate={2017-01-14T18:30:00.000Z}
  />
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hidePrevButton={true}
    minDate={2017-01-31T18:29:59.999Z}
    month={2017-02-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    startDate={2017-01-14T18:30:00.000Z}
  />
</div>
`;

exports[`<DateRangePickerControl/> renders consecutive months with dropdowns 1`] = `
<div
  className="rdr-range-control-wrapper"
>
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hideNextButton={true}
    month={2017-01-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    showDropdowns={true}
    startDate={2017-01-14T18:30:00.000Z}
  />
  <CalendarMonth
    endDate={2017-03-13T18:30:00.000Z}
    hidePrevButton={true}
    minDate={2017-01-31T18:29:59.999Z}
    month={2017-02-14T18:30:00.000Z}
    onDayClick={[Function]}
    onDayHover={[Function]}
    onMonthChange={[Function]}
    onNextClick={[Function]}
    onPrevClick={[Function]}
    onYearChange={[Function]}
    showDropdowns={true}
    startDate={2017-01-14T18:30:00.000Z}
  />
</div>
`;
